syft/pkg/cataloger/java/parse_pom_xml.go

```diff
@@ -53,59 +47,172 @@ func (p pomXMLCataloger) Catalog(ctx context.Context, fileResolver file.Resolver
+	for _, pom := range poms {
+		location := pomLocations[pom] // should always exist
+
+		id := r.ResolveID(ctx, pom)
+		mainPkg := newPackageFromMavenPom(ctx, r, pom, location)
+		if mainPkg == nil {
+			continue
+		}
+		resolved[id] = mainPkg
+		pkgs = append(pkgs, *mainPkg)
+	}
+
+	// catalog all dependencies
 	for _, pom := range poms {
-		pkgs = append(pkgs, processPomXML(ctx, r, pom, r.pomLocations[pom])...)
+		location := pomLocations[pom] // should always exist
+
+		id := r.ResolveID(ctx, pom)
+		mainPkg := resolved[id]
+
+		newPkgs, newRelationships, newErrs := collectDependencies(ctx, r, resolved, mainPkg, pom, location, p.cfg.ResolveTransitiveDependencies)
+		pkgs = append(pkgs, newPkgs...)
+		relationships = append(relationships, newRelationships...)
+		errs = unknown.Join(errs, newErrs)
 	}
-	return pkgs, nil, errs
+
+	return pkgs, relationships, errs
 }
 
-func readPomFromLocation(fileResolver file.Resolver, pomLocation file.Location) (*gopom.Project, error) {
+func readPomFromLocation(fileResolver file.Resolver, pomLocation file.Location) (*maven.Project, error) {
 	contents, err := fileResolver.FileContentsByLocation(pomLocation)
 	if err != nil {
 		return nil, err
 	}
 	defer internal.CloseAndLogError(contents, pomLocation.RealPath)
-	return decodePomXML(contents)
+	return maven.ParsePomXML(contents)
+}
+
+// newPackageFromMavenPom processes a single Maven POM for a given parent package, returning only the main package from the pom
+func newPackageFromMavenPom(ctx context.Context, r *maven.Resolver, pom *maven.Project, location file.Location) *pkg.Package {
+	id := r.ResolveID(ctx, pom)
+	parent, err := r.ResolveParent(ctx, pom)
+	if err != nil {
+		// this is expected in many cases, there will be no network access and the maven resolver is unable to
+		// look up information, so we can continue with what little information we have
+		log.Trace("unable to resolve parent due to: %v", err)
+	}
+
+	var javaPomParent *pkg.JavaPomParent
+	if parent != nil { // parent is returned in both cases: when it is resolved or synthesized from the pom.parent info
+		parentID := r.ResolveID(ctx, parent)
+		javaPomParent = &pkg.JavaPomParent{
+			GroupID:    parentID.GroupID,
+			ArtifactID: parentID.ArtifactID,
+			Version:    parentID.Version,
+		}
+	}
+
+	pomLicenses, err := r.ResolveLicenses(ctx, pom)
+	if err != nil {
+		log.Tracef("error resolving licenses: %v", err)
+	}
+	licenses := toPkgLicenses(&location, pomLicenses)
+
+	m := pkg.JavaArchive{
+		PomProject: &pkg.JavaPomProject{
+			Parent:      javaPomParent,
+			GroupID:     id.GroupID,
+			ArtifactID:  id.ArtifactID,
+			Version:     id.Version,
+			Name:        r.ResolveProperty(ctx, pom, pom.Name),
+			Description: r.ResolveProperty(ctx, pom, pom.Description),
+			URL:         r.ResolveProperty(ctx, pom, pom.URL),
+		},
+	}
+
+	p := &pkg.Package{
+		Name:    id.ArtifactID,
+		Version: id.Version,
+		Locations: file.NewLocationSet(
```

Comment:
Is this a bug? When packages are entirely from remote online sources and not directly based on a file, should the source that caused the package to appear get set or should this be empty?

Reply:
this is questionable for sure, but the rule of thumb here is to track the files you needed to open and read to detect the packages existence. In this case it's a little misleading since the pom might be for an entirely different package, however, it still is defensible since without that location there really would be no package to find.

I think in the future we want to be able to have locations that represent remote URLs and not exclusively file paths -- that way we could express more accurate info here on the package as to where it was discovered from.